python/ray/train/mosaic/_mosaic_utils.py

```diff
@@ -0,0 +1,225 @@
+                )
+            result = mosaic_trainer.fit()
+
+            chkpt_dict = result.checkpoint.to_dict()
+
+            in_memory_logger = chkpt_dict["in_memory_logger"]
+            last_checkpoint = chkpt_dict["last_checkpoint"]
+            all_checkpoints = chkpt_dict["all_checkpoints"]
+
+    Args:
+        in_memory_logger: The list of Composer InMemoryLogger that would be used in
+            Composer trainer initialization.
+        checkpoint_saver: A Composer ``CheckpointSaver`` that the callback will wrap.
+            If this argument is provided, then the parent class is initialized with the
+            passed in ``CheckpointSaver`` object's attributes. Otherwise, the parent
+            class is initialized with args provided.
+        args: Arguments for initializing a Composer ``CheckpointSaver`` object.
+    """
+
+    def __init__(
+        self, in_memory_logger, checkpoint_saver: CheckpointSaver = None, **args
+    ):
+        self.in_memory_logger = in_memory_logger
+        self.last_checkpoint = None
+        self.checkpoint_count = 0
+
+        if checkpoint_saver:
+            super(RayTrainReportCallback, self).__init__(
+                checkpoint_saver.folder,
+                checkpoint_saver.filename,
+                checkpoint_saver.artifact_name,
+                checkpoint_saver.latest_filename,
+                checkpoint_saver.latest_artifact_name,
+                checkpoint_saver.save_interval,
+                overwrite=checkpoint_saver.overwrite,
+                num_checkpoints_to_keep=checkpoint_saver.num_checkpoints_to_keep,
+                weights_only=checkpoint_saver.weights_only,
+                **args
+            )
+        else:
+            super(RayTrainReportCallback, self).__init__(**args)
+
+    def close(self, state: State, logger: Logger) -> None:
+        del logger  # unused
+        checkpoint = Checkpoint.from_dict(
+            {
+                "last_checkpoint": self.last_checkpoint,
+                "in_memory_logger": self.in_memory_logger,
+                "all_checkpoints": self.saved_checkpoints,
+            }
+        )
+        session.report(metrics={}, checkpoint=checkpoint)
+
+    def epoch_checkpoint(self, state: State, logger: Logger) -> None:
+        super().epoch_checkpoint(state, logger)
+        self._update_checkpoint(state)
+
+    def batch_checkpoint(self, state: State, logger: Logger) -> None:
+        super().batch_checkpoint(state, logger)
+        self._update_checkpoint(state)
+
+    def _update_checkpoint(self, state: State):
+        # check that the saved checkpoint is not redundant
+        if len(self.saved_checkpoints) > self.checkpoint_count:
+            self.last_checkpoint = [
+                chkpt_path.absolute() for chkpt_path in self.saved_checkpoints[-1][1]
+            ]
+            self.checkpoint_count = len(self.saved_checkpoints)
+            session.report(
+                metrics={},
```

Comment:
We shouldn't report checkpoints separately from metrics - we want them to be associated with those so that we can determine the best checkpoint to keep. We also would like to keep parity between training iterations and the number of reports made. Can we somehow combine metrics and checkpoint reporting here?